File: src/main/scala/s/template_method/inheritance/StringDisplay.scala
/* (C) 2022 */
package s.template_method.inheritance

class StringDisplay(val string: String) extends AbstractDisplay {
  val width: Int = string.length

  override def open(): Unit = {
    printLine()
  }

  override def print(): Unit = {
    println("|" + string + "|")
  }

  override def close(): Unit = {
    printLine()
  }

  private def printLine(): Unit = {
    scala.Predef.print("+")
    for (i <- 0 until width) {
      scala.Predef.print("-")
    }
    scala.Predef.println("+")
  }
}

File: src/main/kotlin/k/template_method/inheritance/StringDisplay.kt
package k.template_method.inheritance

class StringDisplay(private val string: String) : AbstractDisplay() {
    private val width: Int = string.length

    override fun open() {
        printLine()
    }

    override fun print() {
        println("|$string|")
    }

    override fun close() {
        printLine()
    }

    private fun printLine() {
        print("+")
        for (i in 0 until width) {
            print("-")
        }
        println("+")
    }
}
